pkg/controller/elasticsearch/pdb/roles.go

```diff
@@ -0,0 +1,518 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package pdb
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sort"
+
+	appsv1 "k8s.io/api/apps/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	esv1 "github.com/elastic/cloud-on-k8s/v3/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/common/metadata"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/v3/pkg/utils/set"
+)
+
+var (
+	// group the statefulsets by the priority of their roles.
+	// master, data_*, ingest, ml, transform, coordinating, and we ignore remote_cluster_client as it has no impact on availability
+	priority = []string{"master", "data", "data_frozen", "ingest", "ml", "transform", "coordinating"}
+	// All data role variants should be treated as a generic data role for PDB purposes
+	dataRoles = []esv1.NodeRole{
+		esv1.DataRole,
+		esv1.DataHotRole,
+		esv1.DataWarmRole,
+		esv1.DataColdRole,
+		esv1.DataContentRole,
+		// Note: DataFrozenRole is excluded as it has different disruption rules (yellow+ health)
+	}
+)
+
+// normalizeRole returns the normalized form of a role where any data role
+// is normalized to the same data role.
+func normalizeRole(role esv1.NodeRole) esv1.NodeRole {
+	if slices.Contains(dataRoles, role) {
+		return esv1.DataRole
+	}
+	return role
+}
+
+// reconcileRoleSpecificPDBs creates and reconciles PodDisruptionBudgets per nodeSet roles for enterprise-licensed clusters.
+func reconcileRoleSpecificPDBs(
```

Comment:
nit, we are reconciling PDBs here while this file is named `roles.go`, and we have another file named `reconcile.go`.

I wonder if this function should go in `reconcile.go`? Or maybe we should have `reconcile_default.go` and `reconcile_roles_specific.go` (or something along those lines)?

Reply:
I went with `reconcile_default.go` and `reconcile_with_roles.go`.. Lmk if this works for you.